### ISSUE TITLE
Automatically create a job matrix

### ISSUE THREAD

**Issue by Emiller88** [askimed/nf-test#102]
I'd love to be able to automate a job matrix for our nf-core jobs. Kinda similar to https://github.com/askimed/nf-test/issues/38

Right now we're either running all the tests on one runner:
```yaml
      # Run nf-test
      - name: Run nf-test
        run: nf-test test tests/pipeline/ --profile=test,${{ matrix.profile }} --junitxml=test.xml
```
For each container profile

Or we're creating a strategy like:
```yaml
    strategy:
      fail-fast: false
      matrix:
        NXF_VER:
          - "22.10.1"
          - "latest-everything"
        aligner:
          - "bismark"
          - "bismark_hisat"
          - "bwameth"
        profile:
          - "docker"
```

And testing with
```yaml
      - name: Run nf-test
        if: steps.condition.outputs.skip != 'true'
        run: |
          nf-test test tests/pipeline/${{ matrix.aligner }}/ --profile "test,${{ matrix.profile }}" --tap=test.tap
```

So it requires quite a bit of manual work to keep up with the CI testing right now and keep it flowing smoothly.

I'm thinking if we could list all the tags with a `nf-test list tags` then we could pull in all the tags and run those separately by default and then use the GH actions `include:` to get cross-overs of tags or whatever. Then we'd be able to do something like [this](https://github.com/nf-core/demultiplex/blob/21dad4b89a6d6994e660aa158da18f6499d93e83/.github/workflows/ci.yml#L21-L33) to automate the job creation, of one GH job per nf-test tag.

But I guess I could hack something together with `nf-test list` right now to just run every test separately 🤔

**Comment by lukfor** [comment]
Hi Edmund,

Thanks for working on a clean and reusable github action. We appreciate it a lot!

I started a new branch https://github.com/askimed/nf-test/commit/dd531f8a46c69ed6e601361ee55ff396c24badbf and added the `--tags` option to the existing `list` function to print only the available tags to stdout:

```
nf-test list --tags         

🚀 nf-test 0.7.3
https://code.askimed.com/nf-test
(c) 2021 - 2023 Lukas Forer and Sebastian Schoenherr

fastqc
snakemake
```

With the `--silent` option you could also hide the header to get a clean list (maybe easier for further processing):

```
nf-test list --tags --silent
fastqc
snakemake
```

Would this help? Do you prefer an other output format?

**Comment by Emiller88** [comment]
I think that kinda fixes it, I was going to parse it all. If we could get to one test, one job that might be cool too, I know sarek and rnaseq each of the tests take a long time.

While I'm asking what about `--silent` or machine readable:
```console
[tests/functions/demultiplex.nf.test] Test Functions

  [tests/functions/demultiplex.nf.test@31a4c2e5] Test Function extract_csv

[tests/pipeline/fqtk.nf.test] Test Workflow main.nf

  [tests/pipeline/fqtk.nf.test@30dfbce1] FQTK

[tests/pipeline/bclconvert.nf.test] Test Workflow main.nf

  [tests/pipeline/bclconvert.nf.test@48c61014] BCL-CONVERT
```
to instead:

```console
tests/functions/demultiplex.nf.test@31a4c2e5
tests/pipeline/fqtk.nf.test@30dfbce1
tests/pipeline/bclconvert.nf.test@48c61014
```

I'll write a small script for it if not, but it'd probably be cleaner in nf-test proper 😁

**Comment by Emiller88** [comment]
> Should we also add an additional `--json` flag to write the list of tests/tags as json to stdout? I think gh action supports building a matrix based on json. Maybe this is simpler and easier? 🤔 I am not a gh action expert.
This would also be great, whichever is easier, and if you only want to tackle the json that'd be perfect as well!

>     * According the one job per tag approach: nf-test supports multiple tags per test. Such tests will then run multiple times.

Exactly, that's why I switched to using the test names! 😆 rnaseq in the rewrite was running multiple tests.

But you got me thinking again... if you do the json, then we can parse for tags, and then split the jobs up. Because I have a setup that looks for changes in the code, and then only runs the pipeline pathways that have changed in the PR, which we could do based off the tags.

>     * An other much simpler approach would be to distributed the tests accross a fixed number of machines (jest and playwright calls it sharding).  See here for an example: https://playwright.dev/docs/test-sharding

This would also work really well! I have a feeling this is the only way forward with testing pipelines like this, because each test execution by itself takes way too long. I wonder if it could be combo'd with tags... `nf-test test --tag aligner --shard=1/2`

I'd vote in-terms of bang for buck:
1. List tests `--json` option
2. Sharding (which I guess isn't too bad, you just take the tests, and divide them 🤔)

**Comment by lukfor** [comment]
Json output is implemented in this PR https://github.com/askimed/nf-test/pull/105 

Any feedback @Emiller88? Otherwise we will merge it and include it in the next release 👍 

----

> I have a feeling this is the only way forward with testing pipelines like this, because each test execution by itself takes way too long

That's also my feeling and it is imho a simple yet scalable solution. We will implement sharding in an other PR. I like the idea :)

**Comment by lukfor** [comment]
We merged PR #105 👍  I close this issue for now, please create a new one for Sharding.
